### PR TITLE
Field split decomposition bug fixes

### PR DESCRIPTION
**Pull request by lindsayad**
- For `navier_stokes` field splits: Back when this embedding code was written we didn't support vector variables in MOOSE. However, now we do so we have to make sure we get all the dof components of the vector variables
- For `contact` field splits: we were over-providing dofs both for the contact and interior portions of the splits. 
- With the changes in this PR we now satisfy error checking that the the sum of the splits are equal to the system matrix size both for `navier_stokes` and `contact`

Refs #22359

### CODE REVIEW

**Comment by lindsayad** [comment]
@js-jixu I believe this should fix your errors when running field split

**Review by GiudGiud — COMMENTED** [review]
should there be a dof_indices API that fills a vector?

seems that s what we have in the places that did not need modification

**Comment by roystgnr** [comment]
```
 /**
   * Fills the vector \p di with the global degree of freedom indices
   * for the \p node, for one variable \p vn.
   */
  void dof_indices (const Node * const node,
                    std::vector<dof_id_type> & di,
                    const unsigned int vn) const;
```
?

It's a `DofMap` method but we seem to have that handy in the code here.

**Review by roystgnr — DISMISSED** [review]

**Review by GiudGiud — DISMISSED** [review]

**Comment by lindsayad** [comment]
@fdkong @dewenyushu my new assertion causes a failure in the Modules Debug recipe for the contact field split test:
```
contact/test:fieldsplit.2block-3D: Assertion `libMesh::cast_int<libMesh::numeric_index_type>(split_size_sum) == dmm->_nl->nonlinearSolver()->system().get_system_matrix().local_m()' failed
contact/test:fieldsplit.2block-3D: Local split size sum 270 and local system matrix size 162 do not match.
```
Looking at `-snes_view` I see that the IS size of the `fieldsplit_contact_` is 162 which is actually the size of the system matrix. (`fieldsplit_interior_` has IS size of 108) Does this make sense? It doesn't make sense to me... I tend to assume a split will form two sets that don't have an intersection and union to the global index set. But maybe things are different for contact?

**Comment by js-jixu** [comment]
> @js-jixu I believe this should fix your errors when running field split

I'll update my moose after fixing the bug and try it out.👍👍👍

**Comment by lindsayad** [comment]
Ok the error checking is not relevant in profiling 👍

**Comment by lindsayad** [comment]
@recuero pinging you for awareness. I'm not sure how much you've played around with field splits for contact

**Comment by recuero** [comment]
I haven't played with it, but it has the potential to help with scalability so I am interested. I see you re-enabled the mortar test, thanks for the fixes!

File: framework/src/splits/Split.C [hunk 1]
```diff
@@ -49,12 +49,6 @@ Split::validParams()
   params.addParam<MooseEnum>(
       "schur_pre", SchurPreEnum, "Type of Schur complement preconditioner matrix");
 
-  MooseEnum SchurAInvEnum("diag lump", "diag");
-  params.addParam<MooseEnum>(
-      "schur_ainv",
-      SchurAInvEnum,
-      "Type of approximation to inv(A) used when forming S = D - C inv(A) B");
```
Comment (GiudGiud) [inline framework/src/splits/Split.C:56]: what motivates this?

File: modules/contact/test/tests/fieldsplit/tests [hunk 2]
```diff
@@ -11,11 +11,10 @@
...
-    input = 'frictional_mortar_FS.i'
-    exodiff = 'frictional_mortar_FS_out.e'
+    input = 'frictionless_mortar_FS.i'
+    exodiff = 'frictionless_mortar_FS_out.e'
     abs_zero = 1e-8
-    mesh_mode = REPLICATED
-    skip = 'Weighted mortar'
-    requirement = 'The system shall allow split preconditioning of frictional mortar contact.'
+    requirement = 'The system shall allow split preconditioning of frictionless mortar contact.'
+    petsc_version = '>= 3.16.6' # early versions of petsc can't handle change in sparsity pattern in field split
```
Comment (GiudGiud) [inline modules/contact/test/tests/fieldsplit/tests:18]: in my limited understand of contact, frictionless and frictional would be different.
can you document why you are making that change and is it necessary for you to? does frictional not make sense here?

File: framework/src/splits/Split.C (hunk 1 above)
Comment (lindsayad) [reply framework/src/splits/Split.C:56]: The option is never used by PETSc. It's a red herring having the option available to users

File: modules/contact/test/tests/fieldsplit/tests (hunk 2 above)
Comment (GiudGiud) [reply modules/contact/test/tests/fieldsplit/tests:18]: or does it just not matter

Comment (lindsayad) [reply modules/contact/test/tests/fieldsplit/tests:18]: I guess I did it because the node-face version was frictionless

Comment (lindsayad) [reply modules/contact/test/tests/fieldsplit/tests:18]: I can work on changing it back if you'd prefer

Comment (lindsayad) [reply modules/contact/test/tests/fieldsplit/tests:18]: Or I think I would add a frictional test and keep what is now the frictionless test

Comment (lindsayad) [reply modules/contact/test/tests/fieldsplit/tests:18]: I didn't think much about changing the requirement given that the test was being skipped coming into this

Comment (GiudGiud) [reply modules/contact/test/tests/fieldsplit/tests:18]: Well depends on if it matters wrt to the subject at hand

Comment (lindsayad) [reply modules/contact/test/tests/fieldsplit/tests:18]: With mortar contact you don't actually need to use the `ContactSplit` object nor the contact data members on `DM_Moose`. You can just rely on splitting by variables, whereas for node-face you don't have Lagrange multipliers so you need to split based on displacement interior and contact interface dofs, and that *does* use the `ContactSplit` object and the contact data members on `DM_Moose`. So in terms of using the contact lines of code I've touched in this PR, only the node-face test does that.

Comment (lindsayad) [reply modules/contact/test/tests/fieldsplit/tests:18]: It doesn't cost me much to add a frictional test. I'll just do that now

Comment (recuero) [reply modules/contact/test/tests/fieldsplit/tests:18]: Didn't look at the code... Just wanted to add that one may want to use field split with penalty mortar contact, without LMs.

Comment (lindsayad) [reply modules/contact/test/tests/fieldsplit/tests:18]: For which we still have the ContactSplit which we test in the node face test. 

We don’t have penalty mortar mechanical contact do we?

Comment (recuero) [reply modules/contact/test/tests/fieldsplit/tests:18]: We should have penalty mortar contact soon. 

I assume using the condensed LMs approach would make the same usage of field split as the node face test you pointed out too.

**Comment by lindsayad** [comment]
Are we good to go here?

**Comment by GiudGiud** [comment]
I think we ll want a test with vector variables and a test with array variables

**Comment by bwspenc** [comment]
@lindsayad I'm glad to see you're working on this! What would the implications of over-providing dofs been for contact? It was a long time ago when I tried this for contact, but what I remember was that it had good convergence rates, but it was just really slow.

**Comment by lindsayad** [comment]
I think what would happen is that you would precondition the overlapping indices twice ... and essentially whatever preconditioner went last would "win".

**Comment by lindsayad** [comment]
Ok @GiudGiud I've added FSP framework level tests for both vector and array variables

**Comment by lindsayad** [comment]
@GiudGiud or @roystgnr when you have time, I'd appreciate a review on this. @js-jixu has been poking me on this for a bit now and I'd like to help

**Comment by GiudGiud** [comment]
will do tomorrow

**Comment by GiudGiud** [comment]
This is a net and big fix so I m recommending we merge it. 
While testing it I fed it a lot of bad syntax that we should be able to catch. And I found an issue with array variable Schur field splits. 

Here we have 3 variables, u_vec, v_vec, w
**Duplicate variable specifications that matches the size**
```[Preconditioning]
  [FSP]
    type = FSP
    topsplit = 'uv'
    [uv]
      splitting = 'u v w'
      # Generally speaking, there are four types of splitting we could choose
      # <additive,multiplicative,symmetric_multiplicative,schur>
      splitting_type  = additive
    []
    [u]
      vars = 'w'
      petsc_options_iname = '-pc_type -ksp_type'
      petsc_options_value = '     hypre preonly'
    []
    [v]
      vars = 'u_vec w'
      petsc_options_iname = '-pc_type -ksp_type'
      petsc_options_value = '     hypre  preonly'
    []
    [w]
      vars = 'w'
      petsc_options_iname = '-pc_type -ksp_type'
      petsc_options_value = '     hypre  preonly'
    []
  []
[]
```

**Schur dies unless we pass these options**
```
petsc_options_iname = '-pc_fieldsplit_schur_precondition'
petsc_options_value = 'selfp'
```
with
```
[0]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[0]PETSC ERROR: No support for this operation for this object type
[0]PETSC ERROR: Mat type schurcomplement
[0]PETSC ERROR: See https://petsc.org/release/faq/ for trouble shooting.
[0]PETSC ERROR: Petsc Release Version 3.16.6, unknown 
```

I added x_arr an array variable
**Array var Schur field splits involving dirichlet BCs (on the array var or on the other variables) break** 
See attached input
[vector-test.txt](https://github.com/idaholab/moose/files/10147011/vector-test.txt)
This could be fixed in this PR or a follow up

and unrelated
**array diffusion doesnt check the size of the arrays without an assert**
```
Assertion `(*_d_array)[_qp].size() == _var.count()' failed
diffusion_coefficient size is inconsistent with the number of components of array variable
at /Users/giudgl/projects/moose_v5/framework/src/kernels/ArrayDiffusion.C, line 53
```

**Review by GiudGiud — APPROVED** [review]

**Comment by lindsayad** [comment]
Thanks for the investigation and review. Do you want to put those into issues? Otherwise I can

**Comment by GiudGiud** [comment]
I'll do it I have the inputs handy that show each

**Pull request merged**